### ISSUE TITLE
Week 4: Node 1

### ISSUE THREAD

**Issue by bradreeder** [foundersandcoders/master-reference#30]
This issue is intended for usage by mentor's who get assigned to Node core 1 week and for discussion around the week. As a single source for S-G-C notes and FAC9 feedback, discussion, external feedback, and to commit their work against for review.

**Comment by bradreeder** [comment]
FAC9 feedback #18 

* Move streams as a readme topic from week 4 to week 5, as it is advanced and requires core concepts before understanding.

**Comment by bradreeder** [comment]
Week 4 S-G-C notes:

##### Stop

* Readme presentations being so long + in depth
* Stop testing core functionality (don't test the obvious like HTTP module)
* lollipopping individuals (puts people on the spot)
* carefully targeted speakers (Make sure speaker is aware of the level of knowledge of the group)

##### Go

* presentation workshops, feedback on group presentations
* clarify what to test and not to test
* A more applied test workshop
* clarify callbacks in week 3 in advance of week 4.

##### Continue

* step by step workshops like Nodegirls
* explaining new things like you've never heard of them before
* well thought of projects
* node shell workshop!
* Adjusting timings based on our needs without sacrificing the content of the week
* resources few enough to feel manageable
* Present on Monday
* Make event loop video mandatory (maybe watch over weekend in advance?)

**Comment by njsfield** [comment]
Following on from @stevehopkinson and @SavageWilliam's suggestion [here](https://github.com/foundersandcoders/master-reference/issues/31), to map out the overall learning outcomes of both Week 1 & 2, it could be worth building on the past learning outcomes [here](https://foundersandcoders.gitbooks.io/fac9/content/week4/learning-outcomes.html) by mixing in additional learning outcomes that were gained from when Node Week was extended, as well SGC notes. So could the following be added (as Week 2 priorities?)-

- Understanding of more complex UNIX commands (piping/grep/sort etc).
- Understanding the Node Shell.
- Understanding how to create *npm* packages to reuse across projects.
- Understanding the difference between Streams and Buffers in Node, their benefits and how to implement them.
- Understanding Events in Node and using the EventEmitter class.

Any thoughts? With learning outcomes separate we can ensure no crossover of content, but can discuss project ideas to ensure Week 2's project builds on from Week 1's (if appropriate)

**Comment by msachi** [comment]
@njsfield I agree that it would be great to cover some of these concepts in more detail. I'm going though Antony Alicea's Node course now and it's quite inspiring how even tricky topics such as streams, buffers and events can be made very accessible.

If I understand @bradreeder's proposal for the [curriculum](https://github.com/foundersandcoders/master-reference/blob/master/curriculum.md) correctly, however, concepts such as streams and event emitter would be covered in week 1 (which covers the more theoretical sides of Node) and there's lots of other stuff to cover in week 2:

> commit hooks
> istanbul, ci (travis, codecov)
> node debugger
> error handling in node + async error handling
> tdd server
> project management part 3: Pull requests. Across team QA, passing branches
> Securing the back-end (talk / workshop)

Of these, I think testing the back-end is a substantial topic and we should allow enough time for it. It would also be great if we could cover code coverage and continuous integration better than how this was done in our course.

I think if we were to add anything more to week 2 we need to move some of the content above to week 1.

**Comment by marisid** [comment]
@njsfield @msachi as you've mentioned, I think it's important to clarify the learning objectives for the two Node weeks to avoid unnecessary overlapping. @msachi I think I need to do Tony's course as well - thanks for sharing :+1: 

As I understand the [curriculum](https://github.com/foundersandcoders/master-reference/blob/master/curriculum.md), the first week will be more theoretical (which makes sense) and therefore, we could have streams and event emitters presented in week 1 as part of a workshop (e.g. replacing the Node Shell workshop which was very useful but can be moved to week 2). This could be in the form of a long morning challenge on Wed morning and following the readme/ research afternoon on Tue. 

I have to say that @bradreeder did a wonderful job in week 1 :clap: , which I very much enjoyed and it's a matter of refining it now for the next cohort. Based on the structure of the Node week in our course, the first week could look like below (my suggestions below are in italic):

- Monday: Node walkthrough + Node Girls workshop in the afternoon 
  * *Remind people to watch the [What the Heck is the Event Loop Anyway?](https://www.youtube.com/watch?v=8aGhZQkoFbQ) video over the weekend to be better prepared to understand why we call Node.js: an async (non-blocking), single-threaded, event-driven I/O model*
   * *Highlight the principle of abstraction/ separation of concerns/ modularisation and how this manifests itself (even in the project folder structure)*
   * *Make sure that the Node Girls workshop (or a similar one we decide on) is completed by everyone because I remember quite a few people didn't finish it and they missed the last bits of the exercise, which were quite useful*
- Tuesday:  Tape testing workshop and practice + research in late afternoon
  * *Highlight the nature/ meaning of tests in parallel with introducing how tape works*
- Wednesday: Streams/ buffers/ event emitters walkthrough + workshop, start of team projects
- Thursday: Short morning challenge (e.g. similarly to the http request we did with @des-des in week 8?), team projects and team retrospectives
- Friday: Team projects, SGC, presentations

The selection of the project would require more thinking if we want the second week to build on the first one and maybe the teams continue on the same project for two weeks to have a sense of continuity?

Obviously, this is high level summary of the week but just thought to put it out there for suggestions and feedback, so that we can build on it. @stevehopkinson  and @SavageWilliam what do you think?

**Comment by njsfield** [comment]
Apologies @msachi- I was basing *Week 2*'s content on our structures and SGC feedback…
I was concerned too that topics like Streams and Buffers could be quite overloading in Week 1 (as a lot of people felt) so introducing them in the following week would be more suitable... but as they are theoretical topics it makes sense to cover them as research topics in the first week, with workshops on them to follow…

Which is what @marisd has suggested! :) 
The structure looks great :) 

If the project was **autocomplete** for example (as suggested in the [curriculum](https://github.com/foundersandcoders/master-reference/blob/master/curriculum.md)) we could perhaps ensure that *Week 2*’s project built from *Week 1*’s by adding layers of complexity that are relevant to *Week 2*’s curriculum, e.g.- 
- With TDD- add feature X to your **autocomplete** app (ensure errors are handled in tests)
- With TDD- add feature Y to your **autocomplete** app (ensure errors are handled in tests)
- With TDD- add feature Z to your **autocomplete** app (ensure errors are handled in tests)

etc then...
- Add coverage reports with Istanbul (aim for 100% coverage)
- Sync tests/builds with Travis, send reports to Codecov and display badges on your repo Readme etc

**Comment by bradreeder** [comment]
I intend to not be too vocal here overall as I want you to feel like you have ownership over this. Don't feel obligated to use my repo's like node shell because I'm reviewing, only if it fits your structure. Few points:

* If you go with the same structure for day 1 -- i.e. code-along + node girls --  bare in mind business development is currently 16:00 - 18:00 #27 . You all did node girls for 3 1/2 hours if I recall. If it disrupts your schedule too much we'll see if we can negotiate timings as node day 1 is important.
* @msachi the list you highlighted for week 2 is a reflection of lots of people placing where they thought the most appropriate place in the course was for certain topics. I agree that server-side tdd, continuous integration, code coverage were considered most important on that list. If you decide to mix that with some advanced topics that's fine by me personally. You should focus on coming up with two weeks that are progressive and coherent. Basics then advanced, testing with tape + shot then tdd, etc. I wouldn't want you to cram because you thought it was expected if it will lead to a confusing experience for the student.

I'd bare in mind you don't have to dedicate whole workshops to each topic, sometimes it's possible to cover multiple in one go. I recall covering modularisation in about 15 minutes with you all, giving you an example repo, and you were ready to go.

**Comment by des-des** [comment]
@nogainbar @Jbarget I have you down to teach this week in Nazareth.

**Comment by Jbarget** [comment]
@des-des am i right in thinking you mean "Week 1" not "Node week 1"

@bradreeder @des-des can i suggest a format in the issue titles for issues directly related to the learning weeks in the course. [Week Number][Name of week]

**Comment by des-des** [comment]
derp

**Comment by SavageWilliam** [comment]
# Node Resources

* [Learn Node](https://github.com/nikhilaravi/learn-node) *Week1-Mon-am*
   - npm
   - core modules
   - server logic

* [node-intro-workshop](https://github.com/bradreeder/Node-Intro-Workshop): *Week1-Mon-am*
   - this is a version of 'Learn Node'.

* [Node-Girls](https://github.com/node-girls/workshop-cms): *Week1-Mon-pm*
   - we need to review this and fix the bug/come up with a similar yet alternative workshop.


* [Tape-testing](https://github.com/matthewglover/tape-testing): *Week1-Tue-am*
   - include shot?

* [modularise](https://github.com/sofer/sssk): *Week1-Wed-am*
   - dans repo - can be used to make the web morning challenge

* [node-shell](https://github.com/bradreeder/Node-Shell-Workshop/blob/master/LESSON1.md): *Week2-Tue-am*
   - bradleys wonderful node shell workshop for week 2 tuesday!

**Comment by SavageWilliam** [comment]
# Learning Outcomes

## NODE 1
- What is node
- Node / Core Modules / package.json
- Module Patterns
- Tape
- NPM
- Linting (????)
- Making a server
- Tape

## NODE 2
- Continuous Integration
- Node Debugger
- Error Handling / async Error Handling
- TDD server - introduce shot
- Securing the back end
- Stream / Buffers
- Make Requests from Node

# Projects

## NODE 1
Autocomplete

## NODE 2
a news feed project, 3 options

# Structure

## Node 1
- Weekend
  - Node installation (or come in early on monday)
  - Event loop video
  - ES6 resource
- Mon
  - 10:00 - 13:00
    - Intro to servers and Node
    - Code Along
  - 14:00 - 16:00
    - Tailored Node Girls workshop (without CMS)
- Tue
  - 10:00 - 11:00
    - Modularisation morning challenge)
  - 11:00 - 12:45
    - Tape Workshop
  - 12:45 - 13:00
    - Intro to readme topics + teams
  - 14:00 - 18:00
    - Readmes
      - npm modules (nodemon & linting)
      - Heroku & Env2
      - Core modules
      - module.exports & request/response objects
- Wed
  - 10:00 - 11:00
- Thu
- Fri

## Node 2
- Weekend
  - Event loop video
- Mon
  - 10:00 - 13:00
    - ES6 Morning Challenge
  - 14:00 - 16:00
    - Error handling workshop & testing for errors
- Tue
  - 10:00 - 13:00
    - Node shell work (streams, buffers, events)
  - 14:00 - 18:00
    - Readmes
      - Code Coverage
      - Continuous Integration
      - Securing the backend

- Wed
  - 10:00 - 11:00
    - Modularisation morning challenge
- Thu
- Fri

**Comment by bradreeder** [comment]
Truly smashing, nice work! Main review:

* Not saying you have to include these if it will mess up your week, just so they're listed as missing for week 2: debugging, [commit hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks), `project management part 3: Pull requests. Across team QA, passing branches`, could make `Securing the back-end (talk / workshop)` the external speaker slot?
* Bare in mind when you did the autocomplete project you were given until Monday the following week. You may need to mitigate that they have less time by providing them more guidance than I gave you.
* dwyl resources you might make use of or point students to: [learn-tape](https://github.com/dwyl/learn-tape), [learn-tdd](https://github.com/dwyl/learn-tdd), [learn-istanbul](https://github.com/dwyl/learn-istanbul), [learn-travis](https://github.com/dwyl/learn-travis).

If you do use node-shell in week 2 feel free to fork my repo and modify it to your needs (in return for a star ⭐️  😉 -- do star repo's you use and share the ❤️ ) or raise an issue and make a PR.

**Comment by jsms90** [comment]
@marisid Just a note based on:
> Remind people to watch the What the Heck is the Event Loop Anyway? video over the weekend to be better prepared to understand why we call Node.js: an async (non-blocking), single-threaded, event-driven I/O model

This video is now in their [precourse material](https://github.com/foundersandcoders/master-reference/blob/master/coursebook/precourse/resources.md). But many of us, including @lucymonie, can attest to how much extra you get out of it by rewatching it at many stages during F&C. I think that there is real validity in suggesting that they rewatch it at _this particular point_ in the curriculum.

I just want you guys to be fully aware of what we have already asked of them, so that they don't feel like there is a lack of communication between us :+1:

**Comment by marisid** [comment]
@stevehopkinson @SavageWilliam 
Just for reference, there is a new branch called `week-4-node-1` and Will and I just initialised the branch. Let's keep working off this one for all tasks within Node Week 1 in separate branches. When ready, we can create PRs to merge into our main `week-4-node-1` and ultimately merge into `master` in the end. Hope it's clear :smile:

**Comment by marisid** [comment]
**GitBook responsibilities**

- week schedule @SavageWilliam :heavy_check_mark: 
- short summary & context on the week's theme @marisid
- learning-outcomes @marisid :heavy_check_mark: 
- project @SavageWilliam (PR in progress)
- research afternoon @stevehopkinson 
- resources 

**Workshop responsibilities**

Make your own repo or fork existing workshops and tailor them. Once ready, add links to the schedule (*clarification: it has been agreed that our own resources will go into the `README.md` and any external resources will go into `resources.md`*).

- Intro to servers and Node (in conjunction with the `summary.md`) @marisid
- Code Along (Learn Node/Node-intro-ws) @marisid
- Tailored Node Girls workshop (without CMS)
- ES6 morning challenge
- Tape-testing Workshop (& nodemon)
- Modularisation morning challenge - see Dan's repo

@stevehopkinson, @SavageWilliam we haven't assigned `resources.md` from Gitbook responsibilities and the workshop responsibilities. Let's start assigning ourselves to them by editing this comment. I would like to get the first two from the workshop responsibilities - let me know if you are ok with this and pick some for yourselves. :smile:

**Comment by JMurphyWeb** [comment]
@SavageWilliam I am intrigued to hear about this tape testing work shop.. Is there already something in place? I was wondering if it would be worthwhile to develop this repo into a proper morning challenge/workshop? https://github.com/JMurphyWeb/injector

It would basically get the young FACers to build their own version of shot for testing their new shiny servers.

**Comment by msachi** [comment]
@JMurphyWeb Nice idea! We're teaching them Shot in Node week 2, so it might be worth thinking about what order it would be best to present this. @njsfield

**Comment by JMurphyWeb** [comment]
@msachi or whether we present it at all.. It was originally intended as an alternative to teaching them shot in node week. Its also quite nice that they can build their own tools so things like shot don't seem magic at all when they do come to use them. 

It could also be deployed to npm so that they use/improve their module in the weeks leading up to hapi.

I am happy to develop this into a proper morning challenge/workshop ideally with input from you guys, if there is a desire to have it in the curriculum

**Comment by SavageWilliam** [comment]
@JMurphyWeb Sorry I missed your post. Fantastic idea, which definitely fits with the general mind set of FAC 'less frameworks/understand from first principles'! It should also fit nicely with the time slot of roughly 1hour 45mins. When they finally discover shot, they should then have a much better understanding of what its doing. 

As you mention in the repo deploying to npm could be an ideal stretch goal if they finish this quickly? It will be the Tues of the first intro to node... potentially a lot to take in regarding npm deployment, what do you think? @msachi - I believe you're doing some 'bradley-type' node shell workshops in node 2?

**Comment by msachi** [comment]
@SavageWilliam I agree that this fits into Node week 2 better. @njsfield is doing a server TDD / Shot workshop 10-13 on Monday of Node week 2, not sure if there's still time to fit it in that slot? Or maybe incorporate or at least reference it somehow.

**Comment by SavageWilliam** [comment]
@msachi  Cool, I'll keep my workshop to a general introduction to tape. My Yoti project sprint has been pushed back a week so I'll have time next week to work on these commitments 😄

**Comment by SavageWilliam** [comment]
@marisid @stevehopkinson @msachi. Looking at [Matt's tape tesing WS](https://github.com/matthewglover/tape-testing) (that we received) it looks pretty good for what we want to do? Without reinventing the wheel, any suggestions on improvements? otherwise I suggest we use it!!  

I believe we want to introduce nodemon here ... suitable time in step 7 of the WS?

**Comment by stevehopkinson** [comment]
@SavageWilliam Looks good to me - covers most bases. Would that also be a good point to talk about how to write testable code (i.e. breaking your code up into small, pure functions)?

**Comment by bradreeder** [comment]
Week delivered. Thanks everyone!
